Add generic overload of Implements

thomaslevesque opened the pull request:
Fixes #470

thomaslevesque commented:
Which checks have failed? I don't have access to TeamCity

blairconrad commented:
@thomaslevesque, access to TeamCity is available to all. Follow the "Details" link and log in as guest.
It will tell you… a thing that I don't really understand, but may indicate that albacore was updated on the server and our rake script is no longer happy.

blairconrad commented:
We may need to resurrect #401. @adamralph, am I misreading the situation.

adamralph commented:
@blairconrad indeed, that is exactly what has happened. I've had to fix other projects to get around it by using bundler.

I'l send a PR shortly which fixes it.

adamralph commented:
This *should* fix it https://github.com/FakeItEasy/FakeItEasy/pull/473

blairconrad commented on Source/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs:
```diff
@@ -126,6 +126,11 @@ public IFakeOptionsBuilder<T> Implements(Type interfaceType)
 
+            public IFakeOptionsBuilder<T> Implements<TInterface>()
+            {
+                return Implements(typeof (TInterface));
```
1. nice typeparam name
2. this is optional, and I wouldn't bother commenting if I weren't already going to ask for changes, but the FakeItEasy convention is to start in-class method calls with `this.`. There's a warning omitted by the build, and the owners often take care of these kinds of changes, but you're keen, so I figure I'd push my luck and see how you felt about making the change anyhow.

blairconrad replied:
Oh, an this is even more minor, but an additional warning from the compiler asks for `typeof(TInterface` instead of `typeof (TInterface`. Again, this is a "nice-to-have".

thomaslevesque replied:
2. OK
3. Do you mean without the space between `typeof` and `(`? I think it was added automatically by VS, but I can remove it if you want

blairconrad replied:
You're very accommodating.

Yeah, for number 3, it was the space that was the problem. I'd lost my backticks when copying my comment and didn't notice.
Not sure about VS adding it. For me it doesn't, but that could be due to my VS configuration or the fact that I have ReSharper (and its configuration, I suppose). Dealing with various projects' conventions is a pain, no?
The warnings emitted by the compilation can be your guide—FakeItEasy usually builds without warnings.

thomaslevesque replied:
I also use ReSharper ; probably a different configuration.

 

But I was not getting any compiler warning… the compiler doesn’t care one bit about whitespace. Are you sure the warning you were seeing was from the compiler?

 

From: Blair Conrad [mailto:notifications@github.com] 
Sent: dimanche 29 mars 2015 01:33
To: FakeItEasy/FakeItEasy
Cc: Thomas Levesque
Subject: Re: [FakeItEasy] Add generic overload of Implements (#471)

 

In Source/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs <https://github.com/FakeItEasy/FakeItEasy/pull/471#discussion_r27350132> :

> @@ -126,6 +126,11 @@ public IFakeOptionsBuilder<T> Implements(Type interfaceType)
>                  return this;
>              }
>  
> +            public IFakeOptionsBuilder<T> Implements<TInterface>()
> +            {
> +                return Implements(typeof (TInterface));

You're very accommodating.

Yeah, for number 3, it was the space that was the problem. I'd lost my backticks when copying my comment and didn't notice.
Not sure about VS adding it. For me it doesn't, but that could be due to my VS configuration or the fact that I have ReSharper (and its configuration, I suppose). Dealing with various projects' conventions is a pain, no?
The warnings emitted by the compilation can be your guide—FakeItEasy usually builds without warnings.

—
Reply to this email directly or view it on GitHub <https://github.com/FakeItEasy/FakeItEasy/pull/471/files#r27350132> .  <https://github.com/notifications/beacon/ABxyWVR4lVrYebo74i9Fj86Tsm_ojlv_ks5n5z-jgaJpZM4D16jk.gif>

blairconrad replied:
Sorry, technically StyleCop warnings. I am not sophisticated, and am sometimes linguistically lazy, so called them "compiler warnings".

I see them in the IDE and the rake build:

![image](https://cloud.githubusercontent.com/assets/3275797/6883641/475b2756-d58d-11e4-968e-8988178985e1.png)

![image](https://cloud.githubusercontent.com/assets/3275797/6883647/6b0936ca-d58d-11e4-8140-48f613f269ec.png)

And as for ReSharper, I do have 

```xml
<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/SPACE_BEFORE_TYPEOF_PARENTHESES/@EntryValue">False</s:Boolean>
```

in my `FakeItEasy.sln.DotSettings.user` file, so that probably explains the difference in behaviour.

thomaslevesque replied:
Oh, I see. I don’t have StyleCop, which explains why I wasn’t getting the warnings

 

From: Blair Conrad [mailto:notifications@github.com] 
Sent: dimanche 29 mars 2015 03:03
To: FakeItEasy/FakeItEasy
Cc: Thomas Levesque
Subject: Re: [FakeItEasy] Add generic overload of Implements (#471)

 

In Source/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs <https://github.com/FakeItEasy/FakeItEasy/pull/471#discussion_r27350332> :

> @@ -126,6 +126,11 @@ public IFakeOptionsBuilder<T> Implements(Type interfaceType)
>                  return this;
>              }
>  
> +            public IFakeOptionsBuilder<T> Implements<TInterface>()
> +            {
> +                return Implements(typeof (TInterface));

Sorry, technically StyleCop warnings. I am not sophisticated, and am sometimes linguistically lazy, so called them "compiler warnings".

I see them in the IDE and the rake build:

 <https://cloud.githubusercontent.com/assets/3275797/6883641/475b2756-d58d-11e4-968e-8988178985e1.png> 

 <https://cloud.githubusercontent.com/assets/3275797/6883647/6b0936ca-d58d-11e4-8140-48f613f269ec.png> 

And as for ReSharper, I do have 

<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/SPACE_BEFORE_TYPEOF_PARENTHESES/@EntryValue">False</s:Boolean>

in my FakeItEasy.sln.DotSettings.user file, so that probably explains the difference in behaviour.

—
Reply to this email directly or view it on GitHub <https://github.com/FakeItEasy/FakeItEasy/pull/471/files#r27350332> .  <https://github.com/notifications/beacon/ABxyWVquKFJi8FxKHBeumftwqZDZV0MRks5n50bNgaJpZM4D16jk.gif>

blairconrad replied:
Hmm. It lives in the `Source\packages`. I thought the warnings would show up anyhow. Huh.
Honestly, I don't really understand the thing. It was in the FakeItEasy when I showed up, and I've never had any experience with it elsewhere.

adamralph replied:
@thomaslevesque the StyleCop.MSBuild package injects build targets into the
csproj which make StyleCop run every single time you build. You don't need
StyleCop installed on your machine and it doesn't matter if you are in VS
or not.

Anyway,  thats just the nitty gritty of the 'how' not the 'why'. StyleCop
is only there to give us *warnings* when code doesn't meet our guidelines
and a few violations here and there certainly won't stop us accepting a PR.
Of course, we'd prefer PR's to be StyleCop compliant but ultimately we as
owners own the burden of keeping the violation count at  or near to zero.
We don't want this to get in the way of contribution.

On 03:09, Sun, 29 Mar 2015 Blair Conrad <notifications@github.com> wrote:

> In Source/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs
> <https://github.com/FakeItEasy/FakeItEasy/pull/471#discussion_r27350377>:
>
> > @@ -126,6 +126,11 @@ public IFakeOptionsBuilder<T> Implements(Type interfaceType)
> >                  return this;
> >              }
> >
> > +            public IFakeOptionsBuilder<T> Implements<TInterface>()
> > +            {
> > +                return Implements(typeof (TInterface));
>
> Hmm. It lives in the Source\packages. I thought the warnings would show
> up anyhow. Huh.
> Honestly, I don't really understand the thing. It was in the FakeItEasy
> when I showed up, and I've never had any experience with it elsewhere.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FakeItEasy/FakeItEasy/pull/471/files#r27350377>.
>

blairconrad replied:
@adamralph, thanks for commenting. That matched my understanding—that contributors don't need to do anything to see the warnings.

And yeah, I hope I didn't give the impression that accepting the PR was contingent on @thomaslevesque fixing the warnings. I was really after the tests, and figured that were I the contributor, I'd've preferred the project owner to ask me for the warnings to be fixed than to learn later that they'd had to go in and fix them.

blairconrad replied:
@adamralph, thanks for commenting. That matched my understanding—that contributors don't need to do anything to see the warnings.

And yeah, I hope I didn't give the impression that accepting the PR was contingent on @thomaslevesque fixing the warnings. I was really after the tests, and figured that were I the contributor, I'd've preferred the project owner to ask me for the warnings to be fixed than to learn later that they'd had to go in and fix them.

blairconrad commented:
Thanks for the PR, @thomaslevesque. If the issue is accepted, one thing I'd ask for would be a test (or tests) for the new functionality. As you can see in the [Contributing.md](https://github.com/FakeItEasy/FakeItEasy/blob/master/CONTRIBUTING.md#tests), we do expect new features to have accompanying specs, and encourage high test coverage.
In this case, I'd be more than happy (would probably prefer it, really) to have new tests added to `FakeItEasy.IntegrationTests.GeneralTests`, since

1. That's where the existing `Implements` is tested.
1. As you point out, this isn't really a new feature, just an overload for an existing feature.
1. There are no existing specs for the fake options builder manipulation. I'm adding some in #465, but there's no sense in us having conflicts. On the other hand, if you decide to add the tests _after_ that PR is merged, adding the Specs would be great. Of course if that happens, it may be necessary to rebase your branch, but we'll know then.

Note that this request doesn't indicate a lack of confidence in the quality of the change. It's just the way we do things. And to guard against (admittedly unlikely) regressions.

blairconrad commented on Source/FakeItEasy/Creation/IFakeOptionsBuilder.cs:
```diff
@@ -66,6 +66,16 @@ public interface IFakeOptionsBuilder<T>
+        /// <returns>Options object.</returns>
+        /// <exception cref="ArgumentException">The specified type is not an interface.</exception>
+        [SuppressMessage("Microsoft.Naming", "CA1716:IdentifiersShouldNotMatchKeywords", MessageId = "Implements", Justification = "Would be a breaking change, might be changed in a later major version.")]
+        IFakeOptionsBuilder<T> Implements<TInterface>();
```
Another optional warning, if you're inclined to fix these kinds of things. The Code Analysis complains about the unused `TInterface`. A suppression like the one on `IAnyCallConfigurationWithNoReturnTypeSpecified.IAnyCallConfigurationWithReturnTypeSpecified` would clean this up:

`[SuppressMessage("Microsoft.Design", "CA1004:GenericMethodsShouldProvideTypeParameter", Justification = "Used to provide a strongly typed fluent API.")]`

or similar.

thomaslevesque replied:
OK

blairconrad commented:
Thanks again. As noted, the "line notes" I made are all optional, and I wouldn't have bothered making them if I weren't already asking for tests.

thomaslevesque commented:
> If the issue is accepted, one thing I'd ask for would be a test (or tests) for the new functionality

Added 2 tests:

- Should_be_able_to_generate_class_fake_that_implements_additional_interface_using_generic
- Should_be_able_to_generate_interface_fake_that_implements_additional_interface_using_generic

(I just copy/pasted the existing tests for this feature and changed them to use the generic `Implements` overload)

thomaslevesque commented:
I pushed all the requested changes

blairconrad commented:
> (I just copy/pasted the existing tests for this feature and changed them to use the generic Implements overload)
(That's probably what I would've done too!)

blairconrad commented:
> I pushed all the requested changes

And they look great. Thanks for that. Nothing to do now but wait!
I suggest you go to sleep while we wait for the other owners. I'm in the east-middle of Canada and think it's nearly bedtime. :smile:  :zzz:

adamralph commented:
Great work @thomaslevesque and @blairconrad!

adamralph commented:
@thomaslevesque look out for your name in the release notes! :trophy:

thomaslevesque commented:
> @thomaslevesque look out for your name in the release notes! :trophy:

lol, I hope I'll be able to contribute in a more significant way in the future ;)

btw, what is the expected release date for 2.0 ?

blairconrad commented:
@thomaslevesque, we'd welcome your contributions. It was a pleasure.

Regarding the release date, you know as much as we do. Well, nearly.

It's all about clearing the [2.0.0 backlog](https://github.com/FakeItEasy/FakeItEasy/issues?q=is%3Aopen+is%3Aissue+milestone%3A2.0.0), which is quite short now, although I'm preparing one additional issue.
Unfortunately, most of the team members are busy with other projects (and I assume lives), so PR reviews are backed up a little. Personally, I'm keen to see it come out soon.

Keep in mind that if you are looking for early access, the 2.0.0 pre-releases are available on NuGet, so you can start using them now. _Nearly_ everything in master has been released.
